text/trace/0192-messaging-semantic-conventions-spec.md

```diff
@@ -0,0 +1,628 @@
+refers to. However, a destination name should only be added to the span name
+when it is of low cardinality. This is usually the case when the destination
+name is a meaningful and manually configured name (like a manually configure
+queue or topic name), it is usually not the case if the destination name is an
+auto-generated identifier (like a conversation id or an auto-generated name for
+an anonymous destination).
+
+> The span name SHOULD consist of the name of the messaging system followed by
+> an [operation name](#operation-name). The destination name MAY be appended if
+> it is of low cardinality.
+
+##### Examples
+
+* `kafka publish shop.orders`
+* `rabbitmq receive print_jobs`
+* `AmazonSQS deliver`
+* `activemq settle`
+
+#### Operation name
+
+The following operations related to messages are covered by these semantic
+conventions:
+
+| Operation name | Description |
+|----------------|-------------|
+| `publish`      | One ore more messages are provided for publishing to an intermediary. |
+| `create`       | A message is created. |
+| `receive`      | One or more messages are requested by a consumer. |
+| `deliver`      | One or more message are passed to a consumer. |
+| `settle`       | One or more message are settled. |
+
+For further details about each of those operations refer, to the [section about trace structure](#trace-structure).
+
+#### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER`, if no `create` spans are present. `INTERNAL` otherwise. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+| `settle`       | `INTERNAL` |
+
+Setting span kinds according to this table ensures, that span links between
+consumer and producers always go from a `PRODUCER` span on the producer side to
+a `CONSUMER` span on the consumer side. This allows analysis tools to interpret
+linked traces without the need of additional semantic hints.
+
+#### Attributes
+
+Attribute             | Type   | Required
+----------------------|--------|---------
+[`messaging.broker`](#messagingbroker) | string | Yes
+[`messaging.operation`](#messagingoperation) | string | Yes
+[`messaging.destination.name`](#messagingdestinationname) | string | For producer spans
+[`messaging.destination.template`](#messagingdestinationtemplate) | string | No
+[`messaging.destination.kind`](#messagingdestinationkind) | string | No
+[`messaging.destination.temporary`](#messagingdestinationtemporary) | string | No
+[`messaging.destination.anonymous`](#messagingdestinationanonymous) | string | No
+[`messaging.source.name`](#messagingsourcename) | string | For consumer spans
+[`messaging.source.template`](#messagingsourcetemplate) | string | No
```

Comment:
I checked the terminology used by different systems on this and found that it's either `wildcard` (NATS, MQTT Clients) or `pattern` (kafka, pulsar). I suggest we stick to existing terminology.

Reply:
+1 for `wildcard`.

Reply:
I'll start off by saying I'm not thrilled with the term `template` myself. But I don't think that `wildcard` is a good choice either, since it is sort of the opposite of I think we should try to achieve with template. `pattern` sounds a bit better, but it is tightly coupled to the wildcard concept, which I hope to show below is not a good match for this attribute. Perhaps `schema` is a better term?

What I think we should be trying to achieve with `template` is to document high cardinality topic levels or fragments. Wildcards express that the consumer wants to receive all values of a topic level or fragment, but it doesn't explain what the topic level is. Why is this useful? I can think of two reasons:
1. It can help with someone viewing traces to either search for related messages (that use the same `template` or `schema`) or to better understand the semantic meaning of the topic in a trace.
2. If a topic has high cardinality topic levels, the topic itself is not suitable for use in a span name. If a `template` or `schema` is available, it would be low cardinality and perhaps we would decide it is useful in a span name. We still have to discuss span names, so I think further discussion of this point should be saved for when we get to the point of discussing span names as a group.

To illustrate these points, suppose you have a topic hierarchy that looks something like this:
`myCompany/vehicles/{vin}/position/{gps-latitude}/{gps-longitude}`

A message might be published with a topic of:
`myCompany/vehicles/4Y1SL65848Z411439/position/45.4236/75.7009`

To someone not familiar with the topic schema being used, this might be hard to decipher. Furthermore, such a field name is not suitable for use within a span name due to its high cardinality. Including the template or schema into the trace helps with both of these problems.

Where the concept of a wildcard fits into the example above is that an MQTT client for example may subscribe to messages like this with a subscription of `myCompany/vehicles/#` or `myCompany/vehicles/+/position/+/+`. In the context of the consumer, the template concept is still useful. The subscription `myCompany/vehicles/#` doesn't say much to someone looking at the trace. However, including the template or schema for the topic allows someone to see the information that the consumer can derive from the topic of the received messages.

For many messaging systems, this concept is not embedded within the messaging system itself, it is only an application layer concern. This means it would be impossible for many auto-instrumented SDK's to include this. However, might it be possible for an application to co-operate with auto-instrumentation and add this attribute to a span started by the SDK?

Regardless, as messaging is becoming more widespread in large organizations, there is an evolving need to maintain a registry of events so that they can be tracked and governance policies can be applied to them. Newer initiatives like AsyncAPI I believe facilitate this with the concept of [channel parameters](https://www.asyncapi.com/docs/reference/specification/v2.0.0#parameterObject). I believe auto-instrumentation added to an AsyncAPI implementation may be able to inject this information automatically.

To sum up, I think this concept is very useful to support. However, coming up with the right name is tricky. Let's discuss further during our next meeting.

Reply:
Agree it's both a 1) important concept for messaging systems that support dynamic topics and 2) a difficult concept name properly.

Reply:
> I checked the terminology used by different systems on this and found that it's either wildcard (NATS, MQTT Clients) or pattern (kafka, pulsar). I suggest we stick to existing terminology.

Currently we know of four different messaging systems supporting this feature (thanks @lmolkova for checking), and there are four different terms for this feature: pattern, wildcard, template, and schema.

Another possible solution would be to make this a messaging system specific attribute. I think this would be justified, as the majority of system do not support it. Kafka would have an attribute `messaging.destination.pattern`, while NATS has `messaging.destination.wildcard`. For a Kafka user who is using a pattern the meaning will be immediately clear, whereas they might be confused by the term "wildcard", "template", or "schema".